In [63]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

<h1><b>B. Khám phá dữ liệu (đan xen với tiền xử lý dữ liệu)</b></h1>

### Đọc dữ liệu từ tập tin (file)

Đầu tiên, đọc dữ liệu từ file "movies.csv" và lưu kết quả vào DataFrame df_movie.

In [64]:
df_movie = pd.read_csv('movies.csv')
df_movie.head()

Title  Rating  Year     Month Certificate Runtime  \
0        Avatar: The Way of Water     7.8  2022  December       PG-13     192   
1  Guillermo del Toro's Pinocchio     7.6  2022  December          PG     117   
2                    Bullet Train     7.3  2022    August           R     127   
3       The Banshees of Inisherin     7.8  2022  November           R     114   
4                           M3gan     6.4  2022  December       PG-13     102   

                            Directors  \
0                       James Cameron   
1  Guillermo del Toro, Mark Gustafson   
2                        David Leitch   
3                     Martin McDonagh   
4                    Gerard Johnstone   

                                               Stars  \
0  Sam Worthington, Zoe Saldana, Sigourney Weaver...   
1  Ewan McGregor, David Bradley, Gregory Mann, Bu...   
2  Brad Pitt, Joey King, Aaron Taylor Johnson, Br...   
3  Colin Farrell, Brendan Gleeson, Kerry Condon, ...   
4  Jenna Davis, Amie Donald, Allison Williams, Vi...   

                        Genre Filming_location         Budget  \
0  Action, Adventure, Fantasy      New Zealand  $350,000,000    
1    Animation, Drama, Family              USA   $35,000,000    
2    Action, Comedy, Thriller            Japan   $85,900,000    
3               Comedy, Drama          Ireland        Unknown   
4    Horror, Sci-Fi, Thriller      New Zealand   $12,000,000    

            Income                       Country_of_origin  
0  $2,267,946,983                            United States  
1        $108,967            United States, Mexico, France  
2    $239,268,602                     Japan, United States  
3     $19,720,823   Ireland, United Kingdom, United States  
4    $171,253,910                            United States

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột

In [65]:
df_movie.shape

(2000, 13)

Dữ liệu có 2000 dòng và 13 cột.

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng trong DataFrame `df_movie` cho biết thông tin của một bộ phim. Có vẻ không có vấn đề các dòng có ý nghĩa khác nhau.

### Dữ liệu có các dòng bị lặp không?

Ta sẽ xem thử rằng trong tập có dòng nào bị trùng không. Nếu trùng thì trả về True, không thì là False và lưu số dòng trùng vào biến num_duplicated_rows

In [66]:
#Tính tổng dòng lặp
num_duplicated_rows = df_movie.duplicated().sum()
is_duplicated = (num_duplicated_rows != 0)
print('So dong lap la {}'.format(num_duplicated_rows))
if is_duplicated == False:
    print('Bo du lieu khong lap')

So dong lap la 0
Bo du lieu khong lap


Vậy không có dòng lặp trong bộ dữ liệu này

### Mỗi cột có ý nghĩa gì?

Ý nghĩa của mỗi cột là:

- **Title** là tên của bộ phim

- **Rating** là điểm đánh giá trên IMDB

- **Year** là năm ra mắt phim.

- **Month** là tháng ra mắt phim.

- **Certificate** là nhãn của bộ phim đó.

- **Runtime** là thời lượng của bộ phim, tính theo phút.

- **Directors** là đạo diễn của bộ phim.

- **Stars** là diễn viên của bộ phim.

- **Genre** là thể loại phim.

- **Filming_location** là địa điểm quay phim.

- **Budget** là chi phí của bộ phim.

- **Income** là doanh thu của bộ phim.

- **Country_of_origin** là nước sản xuất phim.

In [67]:
df_movie.columns

Index(['Title', 'Rating', 'Year', 'Month', 'Certificate', 'Runtime',
       'Directors', 'Stars', 'Genre', 'Filming_location', 'Budget', 'Income',
       'Country_of_origin'],
      dtype='object')

Ta kiểm tra dạng dữ liệu của các cột để xem có đúng với dạng mà ta mong muốn không.

In [68]:
df_movie.dtypes

Title                 object
Rating               float64
Year                   int64
Month                 object
Certificate           object
Runtime               object
Directors             object
Stars                 object
Genre                 object
Filming_location      object
Budget                object
Income                object
Country_of_origin     object
dtype: object

Ta thấy rằng các cột như **Runtime**, **Budge** và **Income** có thể chuyển sang dạng số (numerical) để tiện tính toán. Vậy ta sẽ chuyển các cột này sang dạng số

Đầu tiên là cột **Runtime**, ta sẽ coi các giá trị của nó

In [69]:
df_movie.isna().sum()

Title                 0
Rating                1
Year                  0
Month                 0
Certificate          34
Runtime               0
Directors             0
Stars                 0
Genre                 0
Filming_location      0
Budget                0
Income                0
Country_of_origin     0
dtype: int64

In [70]:
df_movie['Runtime'].unique()

array(['192', '117', '127', '114', '102', '132', '134', '112', '151',
       '107', '188', '105', '131', '125', '126', '98', '101', '161',
       '140', '147', '139', '97', '154', '123', '130', '115', '176', '87',
       '129', '104', '99', '93', '137', '128', '100', '148', '158', '96',
       '86', '108', '124', '89', '111', '103', '135', '95', '119', '136',
       '77', '118', '138', '159', '122', '187', '146', '121', '91', '116',
       '167', '113', '110', '141', '92', '155', '163', '106', '133', '88',
       '150', '156', '242', '109', '143', '152', '83', '120', '90', '144',
       '179', '142', '160', '94', '85', '181', '209', '169', '84', '149',
       '78', '75', '164', '70', '145', '79', '82', '168', '165',
       'Unknown', '153', '180', '50', '157', '172', '80', '162', '170',
       '166', '81', '191', '71', '175', '201', '178'], dtype=object)

Vì có một giá trị là Unknown nên ta sẽ chuyển giá trị này sang NaN bằng cách đặt errors = 'coerce'

In [71]:
df_movie['Runtime'] = pd.to_numeric(df_movie['Runtime'],errors='coerce')
df_movie['Runtime']

0       192.0
1       117.0
2       127.0
3       114.0
4       102.0
        ...  
1995    114.0
1996    117.0
1997     98.0
1998    105.0
1999     79.0
Name: Runtime, Length: 2000, dtype: float64

Tiếp theo là cột **Budget**

In [72]:
df_movie['Budget'].unique()

array(['$350,000,000 ', '$35,000,000 ', '$85,900,000 ', 'Unknown',
       '$12,000,000 ', '$120,000,000 ', '$80,000,000 ', '$20,000,000 ',
       '$10,000,000 ', '$40,000,000 ', '$78,000,000 ', '$1,000,000 ',
       '$16,000,000 ', '$195,000,000 ', '$100,000,000 ', '$250,000,000 ',
       '€ 10,000,000', '$25,000,000 ', '$3,000,000 ', '$150,000,000 ',
       '$170,000,000 ', '$17,000,000 ', '$200,000,000 ', '$4,500,000 ',
       '$32,000,000 ', '$60,000,000 ', '$72,000,000 ', '$2,000,000 ',
       '$68,000,000 ', '$24,000,000 ', '$9,000,000 ', '$50,000,000 ',
       '$90,000,000 ', '$250,000 ', '$85,000,000 ', '$8,000,000 ',
       '₹3,500,000,000', '$22,000,000 ', '€ 7,500,000', '$165,000,000 ',
       'CA$15,000', '$30,000,000 ', '$185,000,000 ', '$75,000,000 ',
       '$18,000,000 ', '$55,000,000 ', '$43,000,000 ', '€ 5,000,000',
       '$300,000,000 ', 'SEK\xa019,000,000', '$190,000,000 ',
       '$15,000,000 ', '$160,000,000 ', '$70,000,000 ', '$24,350,000 ',
       '$820,000 ', '

Ta thấy cột này có đơn vị tiền tệ khác nhau giữa các giá trị, vậy nên ta sẽ chuyển toàn bộ số tiền ở cột này sang đơn vị USD

In [73]:
def convert_price(price):
    
    price = price.replace(' ','')
    price = price.replace(',','')
    
    #Ấn Độ
    if '₹' in price:
        price = price.replace('₹','')
        return int(price)/74
    #Euro
    if '€' in price:
        price = price.replace('€','')
        return int(price)*1.14
    #Bảng Anh
    if '£' in price:
        price = price.replace('£','')
        return int(price)*1.38
    #Canada
    if 'CA$' in price:
        price = price.replace('CA$','')
        return int(price)*0.78
    #Thuỵ Điển
    if "SEK\xa0" in price:
        price = price.replace("SEK\xa0",'')
        return int(price)*0.12
    #Đan Mạch
    if "DKK\xa0" in price:
        price = price.replace("DKK\xa0",'')
        return int(price)*0.15
    #Trung Quốc
    if 'CN¥' in price:
        price = price.replace('CN¥','')
        return int(price)*0.15
    #Nhật Bản
    if '¥' in price:
        price = price.replace('¥','')
        return int(price)/111.68
    #Hàn Quốc
    if '₩' in price:
        price = price.replace('₩','')
        return int(price)/1210
    #Australia
    if 'A$' in price:
        price = price.replace('A$','')
        return int(price)*0.72
    #Nauy
    if 'NOK\xa0' in price:
        price = price.replace('NOK\xa0','')
        return int(price)/9.22
    #USD
    if '$' in price:
        price = price.replace('$','')
        return int(price)
    return price

Chuyển đổi trên toàn bộ cột

In [74]:
df_movie['Budget'] = df_movie['Budget'].apply(convert_price)

Vì có một giá trị là Unknown nên ta sẽ chuyển giá trị này sang NaN bằng cách đặt errors = 'coerce'

In [75]:
df_movie['Budget'] = pd.to_numeric(df_movie['Budget'],errors='coerce')
df_movie['Budget']

0       350000000.0
1        35000000.0
2        85900000.0
3               NaN
4        12000000.0
           ...     
1995            NaN
1996     95000000.0
1997     40000000.0
1998            NaN
1999       500000.0
Name: Budget, Length: 2000, dtype: float64

Cuối cùng là cột **Income**

In [76]:
df_movie['Income'].unique()

array(['$2,267,946,983 ', '$108,967 ', '$239,268,602 ', ...,
       '$141,591,324 ', '$20,235,426 ', '$54,683,487 '], dtype=object)

Có vẻ như cột này chỉ có đơn vị USD nên ta chỉ cần bỏ kí tự đầu

In [77]:
df_movie['Income'] = df_movie['Income'].str.replace(',','')
df_movie['Income'] = df_movie['Income'].str.replace('$','')
df_movie['Income'] = pd.to_numeric(df_movie['Income'],errors='coerce')
df_movie['Income']

C:\Users\HP\AppData\Local\Temp\ipykernel_3268\2057679562.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df_movie['Income'] = df_movie['Income'].str.replace('$','')


0       2.267947e+09
1       1.089670e+05
2       2.392686e+08
3       1.972082e+07
4       1.712539e+08
            ...     
1995    1.062878e+06
1996    1.600992e+08
1997    1.415913e+08
1998    2.023543e+07
1999    5.468349e+07
Name: Income, Length: 2000, dtype: float64

In [78]:
df_movie['Month'].value_counts()

November     260
December     249
October      187
June         166
July         152
September    151
March        150
August       149
January      148
May          144
April        121
February     121
2014           1
2008           1
Name: Month, dtype: int64

In [79]:
df_movie[df_movie['Month'] == '2008']

Title  Rating  Year Month Certificate  Runtime  \
1439  The Lazarus Project     6.1  2008  2008       PG-13    100.0   

       Directors                                              Stars  \
1439  John Glenn  Paul Walker, Piper Perabo, Brooklynn Proulx, B...   

                         Genre Filming_location     Budget  Income  \
1439  Drama, Mystery, Thriller           Canada  9750000.0  9162.0   

                          Country_of_origin  
1439  United States, Canada, United Kingdom

Phim này ra mắt tháng 6

In [80]:
df_movie.loc[1439,'Month'] = 'June'

In [81]:
df_movie[df_movie['Month'] == '2014']

Title  Rating  Year Month Certificate  Runtime      Directors  \
848  Mantus     2.1  2014  2014         NaN      NaN  Enzo Zelocchi   

                                                 Stars  \
848  Enzo Zelocchi, Miryam Negrin, Paul Gregory, Ma...   

                       Genre Filming_location    Budget  Income  \
848  Comedy, Drama, Thriller              USA  250000.0     NaN   

    Country_of_origin  
848     United States

In [82]:
df_movie.loc[848,'Month'] = 'July'

In [83]:
df_movie['Month'].value_counts()

November     260
December     249
October      187
June         167
July         153
September    151
March        150
August       149
January      148
May          144
April        121
February     121
Name: Month, dtype: int64

Phim này ra mắt tháng 7

## XỬ LÝ DỮ LIỆU NAN

Ta xem thử mỗi cột có bao nhiêu dữ liệu nan

In [84]:
df_movie.isnull().sum()

Title                  0
Rating                 1
Year                   0
Month                  0
Certificate           34
Runtime                2
Directors              0
Stars                  0
Genre                  0
Filming_location       0
Budget               304
Income               145
Country_of_origin      0
dtype: int64

Thuộc tính Income và Budget có thể điền bằng cách search trên mạng. Tuy nhiên, việc này là vô cùng tốn thời gian vì dữ liệu thiếu ở hai cột này là khá lớn. Nếu điền giá trị đại diện thì độ chính xác sẽ không cao và cũng có thể dẫn đến hiểu lầm về tập dữ liệu, nên tốt nhất là ta nên bỏ hết những dòng có cột thiếu này. (Bỏ thêm cả Rating nữa vì chỉ có 1 dòng)

In [85]:
df_movie.dropna(subset=['Income','Budget','Rating'],inplace = True)
df_movie.isnull().sum()

Title                0
Rating               0
Year                 0
Month                0
Certificate          5
Runtime              0
Directors            0
Stars                0
Genre                0
Filming_location     0
Budget               0
Income               0
Country_of_origin    0
dtype: int64

Ta xem xét các giá trị của cột nhãn để xem có bất thường gì không.

In [86]:
df_movie['Certificate'].fillna('PG-13',inplace=True)

In [87]:
df_movie['Certificate'].unique()

array(['PG-13', 'PG', 'R', 'Not Rated', 'Unrated', 'TV-MA', 'G', 'NC-17',
       'TV-14'], dtype=object)

Ta thấy các giá trị như Not Rated, Unrated, nan. Đây có thể coi là cùng một loại không xác định được. Nên để an toàn thì ta sẽ chuyển các giá trị này sang Unknown.

In [88]:
df_movie['Certificate'].fillna('Unknown',inplace=True)
df_movie['Certificate'] = df_movie['Certificate'].replace(['Not Rated','Unrated'], 'Unknown')


In [89]:
df_movie['Certificate'].unique()

array(['PG-13', 'PG', 'R', 'Unknown', 'TV-MA', 'G', 'NC-17', 'TV-14'],
      dtype=object)

## Với mỗi cột có kiểu dữ liệu dạng số, các giá trị được phân bố như thế nào?

In [90]:
dtypes = df_movie.dtypes
ind = dtypes[(dtypes == np.float64) | (dtypes == np.int64)].index
ind

Index(['Rating', 'Year', 'Runtime', 'Budget', 'Income'], dtype='object')

Ta có các cột dữ liệu dạng số 'Rating', 'Year', 'Runtime', 'Budget', 'Income'. Ta sẽ tính min, lower quartile, median, upper quartile, max để xem sự phân bố của dữ liệu và có gì bất thường ở các cột này không?

In [91]:
def five_summary(df_movie):
    num_col_info_df = pd.DataFrame([],index= ["min", "lower_quartile", "median", "upper_quartile", "max" ])
    for i in ind:
        data = np.nanpercentile(df_movie[i],(0,25,50,75,100)) #tinh min, lower quartile, median, upper quartile, max
        num_col_info_df[i] = data.round(1)
    return num_col_info_df
five_summary(df_movie)

Rating    Year  Runtime       Budget        Income
min                1.9  2003.0     71.0         12.5  3.492000e+03
lower_quartile     6.2  2007.0    100.0   17250000.0  4.177813e+07
median             6.8  2011.0    112.0   40000000.0  1.058333e+08
upper_quartile     7.3  2016.0    126.0   85000000.0  2.454062e+08
max                9.0  2022.0    209.0  356000000.0  2.922918e+09

Ta thấy ở cột Budget, giá trị bé nhất của nó là 12.5 đô. Điều này khá vô lý vì làm phim thường tốn kha khá tiền.

Ta thử in dòng có Budget bé nhất ra xem sao.

In [92]:
df_movie[df_movie['Budget'] == min(df_movie['Budget'])]

Title  Rating  Year    Month Certificate  Runtime  \
487  Trautmann     7.2  2018  October       PG-13    120.0   

                Directors                                              Stars  \
487  Marcus H Rosenm ller  David Kross, Freya Mavor, John Henshaw, Harry ...   

                         Genre  Filming_location  Budget     Income  \
487  Biography, Drama, Romance  Northern Ireland   12.54  1916514.0   

           Country_of_origin  
487  United Kingdom, Germany

Theo tìm hiểu thì budget của phim này là 12.54 triệu chứ không phải là 12.54 đô. Vậy ta cần sửa lại cho đúng.

In [93]:
df_movie.loc[487,'Budget'] *= 1000000

In [94]:
df_movie.loc[487]

Title                                                        Trautmann
Rating                                                             7.2
Year                                                              2018
Month                                                          October
Certificate                                                      PG-13
Runtime                                                          120.0
Directors                                         Marcus H Rosenm ller
Stars                David Kross, Freya Mavor, John Henshaw, Harry ...
Genre                                        Biography, Drama, Romance
Filming_location                                      Northern Ireland
Budget                                                      12540000.0
Income                                                       1916514.0
Country_of_origin                              United Kingdom, Germany
Name: 487, dtype: object

Vậy Budget đã về đúng giá trị thực

Ta kiểm tra lại các mức độ phân bố

In [95]:
five_summary(df_movie)

Rating    Year  Runtime       Budget        Income
min                1.9  2003.0     71.0       7000.0  3.492000e+03
lower_quartile     6.2  2007.0    100.0   17250000.0  4.177813e+07
median             6.8  2011.0    112.0   40000000.0  1.058333e+08
upper_quartile     7.3  2016.0    126.0   85000000.0  2.454062e+08
max                9.0  2022.0    209.0  356000000.0  2.922918e+09

## Với mỗi cột có kiểu dữ liệu không phải dạng số, các giá trị được phân bố như thế nào?

In [96]:
dtypes = df_movie.dtypes
ind = dtypes[(dtypes != np.float64) & (dtypes != np.int64)].index
ind

Index(['Title', 'Month', 'Certificate', 'Directors', 'Stars', 'Genre',
       'Filming_location', 'Country_of_origin'],
      dtype='object')

Ta có các cột không phải dữ liệu dạng số 'Title', 'Month', 'Certificate', 'Directors', 'Stars', 'Genre', 'Filming_location', 'Country_of_origin'. Các cột 'Title', 'Month', 'Directors', 'Stars', 'Filming_location', 'Country_of_origin' chỉ là các cột tên riêng, không có gì đặc biệt ở các cột dữ liệu này để phân tích.

Ta có một cột categorical 'Certificate' và một cột multichoice 'Genre'. Ta sẽ tính số giá trị khác nhau của các cột này và số lượng xuất hiện của các giá trị đó.

In [97]:
df_movie["Certificate"].value_counts()

R          728
PG-13      663
PG         195
Unknown     32
G           21
TV-MA        7
NC-17        4
TV-14        1
Name: Certificate, dtype: int64

In [98]:
arr = np.concatenate(np.array(df_movie["Genre"].str.split(', ')))
unique_values, value_counts = np.unique(arr, return_counts=True)
for value, count in zip(unique_values, value_counts):
    print(f"{value} : {count}")

Action : 599
Adventure : 484
Animation : 102
Biography : 116
Comedy : 556
Crime : 270
Drama : 770
Family : 97
Fantasy : 181
History : 46
Horror : 199
Music : 37
Musical : 14
Mystery : 177
Romance : 223
Sci-Fi : 200
Sport : 25
Thriller : 289
War : 16
Western : 8


Dữ liệu ở 2 cột này không có gì bất thường.

## Sự phân bố của các dữ liệu dạng category

Ta sẽ coi sự phân bố của các cột: Certificate, Genre, Filming_location, Country_of_origin dựa trên số giá trị khác nhau của chúng và % của các giá trị này

In [99]:
columns_list_cate = np.array(["Certificate","Genre", "Filming_location", "Country_of_origin"])

df_col_cate = pd.DataFrame([],index = columns_list_cate)
num_diff = df_movie[columns_list_cate].nunique()
df_col_cate['Num diff'] = num_diff

diff_ratio = pd.Series()
for i in columns_list_cate:
    v = df_movie[i].value_counts()
    v = round(v*100/(len(df_movie[i])),2)

    diff_ratio.loc[i] = v.to_dict()
df_col_cate['Diff Ratio'] = diff_ratio
df_col_cate

C:\Users\HP\AppData\Local\Temp\ipykernel_3268\1160527616.py:7: FutureWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  diff_ratio = pd.Series()


Num diff                                         Diff Ratio
Certificate               8  {'R': 44.09, 'PG-13': 40.16, 'PG': 11.81, 'Unk...
Genre                   217  {'Action, Adventure, Sci-Fi': 5.57, 'Animation...
Filming_location         95  {'USA': 47.3, 'Canada': 9.69, 'UK': 8.54, 'Unk...
Country_of_origin       351  {'United States': 47.18, 'United Kingdom, Unit...

In [100]:
df_movie.to_csv('clean_movie.csv')